Content.Server/Storage/EntitySystems/EntityStorageSystem.cs

```diff
@@ -46,25 +87,277 @@ private void OnWelded(EntityUid uid, EntityStorageComponent component, WeldableC
+        {
+            if (!ev.ContentsWhitelist.Contains(entity))
+                if (!CanFit(entity, uid))
+                    continue;
+            
+            if (!AddToContents(entity, uid, component))
+                continue;
+
+            count++;
+            if (count >= component.StorageCapacityMax)
+                break;
+        }
+
+        ModifyComponents(uid, component);
+        SoundSystem.Play(component.CloseSound.GetSound(), Filter.Pvs(uid), uid);
+        component.LastInternalOpenAttempt = default;
+        RaiseLocalEvent(uid, new StorageAfterCloseEvent());
+    }
+
+    public bool Insert(EntityUid toInsert, EntityUid container, EntityStorageComponent? component = null)
+    {
+        if (!Resolve(container, ref component))
+            return false;
+
+        if (component.Open)
+        {
+            Transform(toInsert).WorldPosition = Transform(container).WorldPosition;
+            return true;
+        }
+
+        return component.Contents.Insert(toInsert, EntityManager);
+    }
+
+    public bool Remove(EntityUid toRemove, EntityUid container, EntityStorageComponent? component = null)
+    {
+        if (!Resolve(container, ref component))
+            return false;
+
+        return component.Contents.Remove(toRemove, EntityManager);
+    }
+
+    public bool CanInsert(EntityUid container, EntityStorageComponent? component = null)
+    {
+        if (!Resolve(container, ref component))
+            return false;
+
+        if (component.Open)
+            return true;
+
+        if (component.Contents.ContainedEntities.Count >= component.StorageCapacityMax)
+            return false;
+
+        return true;
+    }
+
+    public bool TryOpenStorage(EntityUid user, EntityUid target)
+    {
+        if (!CanOpen(user, target))
+            return false;
+
+        OpenStorage(target);
+        return true;
+    }
+
+    public bool TryCloseStorage(EntityUid target)
+    {
+        if (!CanClose(target))
+        {
+            return false;
+        }
+
+        CloseStorage(target);
+        return true;
+    }
+
+    public bool CanOpen(EntityUid user, EntityUid target, bool silent = false, EntityStorageComponent? component = null)
+    {
+        if (!Resolve(target, ref component))
+            return false;
+
+        if (component.IsWeldedShut)
+        {
+            if (!silent && !component.Contents.Contains(user))
+                _popupSystem.PopupEntity(Loc.GetString("entity-storage-component-welded-shut-message"), target, Filter.Pvs(target));
+
+            return false;
+        }
+
+        //Checks to see if the opening position, if offset, is inside of a wall.
+        if (component.EnteringOffset != (0, 0)) //if the entering position is offset
+        {
+            var targetXform = Transform(target);
+            if (!_interactionSystem.InRangeUnobstructed(target,
+                targetXform.Coordinates.Offset(targetXform.LocalRotation.GetCardinalDir().ToIntVec()),
```

Comment:
This offset should be dependent on `component..EnteringOffset`. Theres probably some nice expression in terms of local coords, but you could also just re-use the offset from earlier (i.e., `targetXform.WorldPosition + targetXform.WorldRotation.RotateVec(component.EnteringOffset)`).